pkg/resource/stack/deployment.go

```diff
@@ -153,12 +172,95 @@ func SerializeDeployment(snap *deploy.Snapshot, sm secrets.Manager, showSecrets
 		}
 	}
 
-	return &apitype.DeploymentV3{
+	deployment := &apitype.DeploymentV3{
 		Manifest:          manifest,
 		Resources:         resources,
 		SecretsProviders:  secretsProvider,
 		PendingOperations: operations,
-	}, nil
+	}
+
+	err := VerifyIntegrity(deployment)
```

Comment:
This is my biggest worry.


Context:

```
// saveSnapshot persists the current snapshot and optionally verifies it afterwards.
func (sm *SnapshotManager) saveSnapshot() error {
	snap := sm.snap()
	if err := snap.NormalizeURNReferences(); err != nil {
		return fmt.Errorf("failed to normalize URN references: %w", err)
	}
	if err := sm.persister.Save(snap); err != nil {
		return fmt.Errorf("failed to save snapshot: %w", err)
	}
	if sm.doVerify {
		if err := snap.VerifyIntegrity(); err != nil {
			return fmt.Errorf("failed to verify snapshot: %w", err)
		}
	}
	return nil
}

func (persister *cloudSnapshotPersister) Save(snapshot *deploy.Snapshot) error {
	token, err := persister.tokenSource.GetToken()
	if err != nil {
		return err
	}
	deployment, err := stack.SerializeDeployment(snapshot, persister.sm, false /* showSecrets */)
	if err != nil {
		return fmt.Errorf("serializing deployment: %w", err)
	}
	return persister.backend.client.PatchUpdateCheckpoint(persister.context, persister.update, deployment, token)
}
```

Before the changes, an error in VerifyIntegrity happens after
`sm.persister.Save` succeeds to persist the snapshot.

After the changes, an error in the equivalent of VerifyIntegrity fails
stack.SerializeDeployment so PatchUpdateCheckpoint does not run.

We were discussing this with @iwahbe the other day and we think that this previous behavior was very intentional - we intended to save the state even if it's not verified, as it might have info about what we did to the cloud (even if it was corrupted) and can assist recovery.

Reply:
Is this change instrumental to what we're doing here with alias rewriting or can it be considered as a separate PR?

Reply:
Perhaps an alternative would be to do `snap.VerifyIntegrity(aliases)` and have that "auto-apply" the alias rewrite as part of verifying. Without changing the control flow.